crates/ra_project_model/src/lib.rs

```diff
@@ -29,12 +29,12 @@ pub enum ProjectWorkspace {
     /// Project workspace was discovered by running `cargo metadata` and `rustc --print sysroot`.
     Cargo { cargo: CargoWorkspace, sysroot: Sysroot },
     /// Project workspace was manually specified using a `rust-project.json` file.
-    Json { project: JsonProject },
+    Json { project: JsonProject, project_location: PathBuf },
 }
 
 impl From<JsonProject> for ProjectWorkspace {
     fn from(project: JsonProject) -> ProjectWorkspace {
-        ProjectWorkspace::Json { project }
+        ProjectWorkspace::Json { project, project_location: PathBuf::new() }
```

Comment:
This feel less then ideal to me. I think we should enforce, at the type level, that the root path is always provided.

Reply:
If `rust-analyzer.linkedProjects` contains rust-project.json inline, we don't have the root path.

Reply:
We still need to behave in a deterministic way when relative paths are specified inline though. I think we should use `rootUri` for `initialize message as a root in this case.

Reply:
Ack. Done. PTAL. I've added a copy of the rootPath to Config.